add support for indexing in `@atomic` macro

kalmarek opened the pull request:
Following the discussion in #54642

Implemented:
- [x] `modifyindex_atomic!`, `swapindex_atomic!`, `replaceindex_atomic!` for `GenericMemory`
- [x] `getindex_atomic`, `setindex_atomic!`, `setindexonce_atomic!` for `GenericMemory` 
- [x] add support for references in `@atomic` macros
- [x] add support for vararg indices in `@atomic` macros 
- [x] tests
- [x] update docstrings with example usage
- ~[ ] update Atomics section of the manual (?)~
- [x] news

@oscardssmith @vtjnash 

# New `@atomic` transformations implemented here:
```julia
julia> @macroexpand (@atomic a[i1,i2])
:(Base.getindex_atomic(a, :sequentially_consistent, i1, i2))

julia> @macroexpand (@atomic order a[i1,i2])
:(Base.getindex_atomic(a, order, i1, i2))

julia> @macroexpand (@atomic a[i1,i2] = 2.0)
:(Base.setindex_atomic!(a, :sequentially_consistent, 2.0, i1, i2))

julia> @macroexpand (@atomic order a[i1,i2] = 2.0)
:(Base.setindex_atomic!(a, order, 2.0, i1, i2))

julia> @macroexpand (@atomicswap a[i1,i2] = 2.0)
:(Base.swapindex_atomic!(a, :sequentially_consistent, 2.0, i1, i2))

julia> @macroexpand (@atomicswap order a[i1,i2] = 2.0)
:(Base.swapindex_atomic!(a, order, 2.0, i1, i2))

julia> @macroexpand (@atomic a[i1,i2] += 2.0)
:((Base.modifyindex_atomic!(a, :sequentially_consistent, +, 2.0, i1, i2))[2])

julia> @macroexpand (@atomic order a[i1,i2] += 2.0)
:((Base.modifyindex_atomic!(a, order, +, 2.0, i1, i2))[2])

julia> @macroexpand (@atomiconce a[i1,i2] = 2.0)
:(Base.setindexonce_atomic!(a, :sequentially_consistent, :sequentially_consistent, 2.0, i1, i2))

julia> @macroexpand (@atomiconce o1 o2 a[i1,i2] = 2.0)
:(Base.setindexonce_atomic!(a, o1, o2, 2.0, i1, i2))

julia> @macroexpand (@atomicreplace a[i1,i2] (2.0=>3.0))
:(Base.replaceindex_atomic!(a, :sequentially_consistent, :sequentially_consistent, 2.0, 3.0, i1, i2))

julia> @macroexpand (@atomicreplace o1 o2 a[i1,i2] (2.0=>3.0))
:(Base.replaceindex_atomic!(a, o1, o2, 2.0, 3.0, i1, i2))

```

kalmarek committed:
add modify, swap, replace for GenericMemory at index

oscardssmith commented on base/genericmemory.jl:
```diff
@@ -207,12 +207,13 @@ getindex(A::Memory, c::Colon) = copy(A)
 ## Indexing: setindex! ##
 
-function setindex!(A::Memory{T}, x, i1::Int) where {T}
+function setindex!(A::GenericMemory{T}, x, i1::Int) where {T}
```
I believe that this was intentionally not implemented (because for almost all algorithms you need a combination of reads and cas rather than reads and writes.

oscardssmith commented:
other than the comment above this looks good! (it is missing an implementation of getindex_atomic which we need because the regular getindex doesn't have a place to pass an order)

kalmarek commented:
@oscardssmith I'm not sure about the semantics of the suggested `getindex_atomic`. One can write a completely generic
```julia
function getindex_atomic(mem::GenericMemory, i::Int, order=default_access_order(mem))
    memref = memoryref(mem, i, @_boundscheck)
    return memoryrefget(memref, i, order, @_boundscheck)
end
```

but there's nothing atomic about the function, except the `order` argument. Is this what you meant?

kalmarek committed:
add atomic getindex, setindex, setindexonce

kalmarek committed:
implement @atomic macros for indexing

oscardssmith commented:
The point is just that (but probably restricted to `AtomicMemory` since
```
julia> m = Memory{Int}(undef, 5);

julia> Base.memoryrefget(memoryref(m, 1), :acquire, false)
ERROR: ConcurrencyViolationError("memoryrefget: non-atomic memory cannot be accessed atomically")
Stacktrace:
 [1] top-level scope
   @ REPL[3]:1
```

The reason we need this function is for the user to be able to provide a stricter order than `:monotonic`

kalmarek committed:
add tests for @atomic indexing

kalmarek commented:
@oscardssmith thanks for the clarification;

Alongside `modifyindex!`, `swapindex!` and `replaceindex!` I've also implemented
* `getindex_atomic(::GenericMemory, i, ...)`
* `setindex_atomic!(::GenericMemory, i, ...)` and
* `setindexonce_atomic!(::GenericMemory, i, ...)`.

I added some converts `(val isa T ? val : convert(T, val)::T)` those `*index` methods and the support for indexing in `@atomic` macro with the following semantics:
```julia
@macroexpand (@atomic a[1])
:(Base.getindex_atomic(a, 1, :sequentially_consistent))

@macroexpand (@atomic order a[1])
:(Base.getindex_atomic(a, 1, order))

@macroexpand (@atomic a[1] = 2.0)
:(Base.setindex_atomic!(a, 1, 2.0, :sequentially_consistent))

@macroexpand (@atomic order a[1] = 2.0)
:(Base.setindex_atomic!(a, 1, 2.0, order))

@macroexpand (@atomicswap a[1] = 2.0)
:(Base.swapindex!(a, 1, 2.0, :sequentially_consistent))

@macroexpand (@atomicswap order a[1] = 2.0)
:(Base.swapindex!(a, 1, 2.0, order))

@macroexpand (@atomic a[1] += 2.0)
:((Base.modifyindex!(a, 1, +, 2.0, :sequentially_consistent))[2])

@macroexpand (@atomic order a[1] += 2.0)
:((Base.modifyindex!(a, 1, +, 2.0, order))[2])

@macroexpand (@atomiconce a[1] = 2.0)
:(Base.setindexonce_atomic!(a, 1, 2.0, :sequentially_consistent, :sequentially_consistent))

@macroexpand (@atomiconce o1 o2 a[1] = 2.0)
:(Base.setindexonce_atomic!(a, 1, 2.0, o1, o2))

@macroexpand (@atomicreplace a[1] (2.0=>3.0))
:(Base.replaceindex!(a, 1, 2.0, 3.0, :sequentially_consistent, :sequentially_consistent))

@macroexpand (@atomicreplace o1 o2 a[1] (2.0=>3.0))
:(Base.replaceindex!(a, 1, 2.0, 3.0, o1, o2))
```

Please have a look and comment!

I'll write/update docstrings once the code is approved.

oscardssmith commented on test/atomics.jl:
```diff
@@ -423,6 +423,69 @@ let a = ARefxy(1, -1)
     @test_throws ConcurrencyViolationError @atomicreplace :monotonic :acquire a.x xchg
 end
 
+let a = AtomicMemory{Float64}(undef, 2)
```
put a `@testset` around this?

oscardssmith replied:
Also we definitely want more tests here. Specifically, testing 32 byte inline structs and a `struct` that has a pointer.

oscardssmith replied:
Also would be good to make sure that the bitsunion case works properly.

oscardssmith replied:
We also should have some "real" tests of this that (for example) test that multi-threaded algorithms actually converge to the correct results.

kalmarek replied:
sure, I can add more tests with e.g. `APadInt` and maybe an index different from 1 ;) I'm not sure though how much more tests are needed.
I see that the primitives (`memoryref*`) are very well tested and `@atomic` macros do nothing more than using them as intended. Adding lots of tests on the higher level doesn't seem to add much value.

I'd suggest a very similar testset for 
1. `AtomicMemory{APadInt}` (non-standard size)
2. `AtomicMemory{Union{Nothing, Int}}` (union type) and
3. `AtomicMemory{Vector{BigInt}}` (non-isbitstype, contains pointer)  

----

What I'd like to test more is the code generated by `@atomic` macro, but I didn't find a straightforward way of comparing e.g. 
```julia
ex = @macroexpand @atomic ...
@test ex == test_ex 
```
without building the `test_ex` by hand. Any tips on this?

As for "real" examples I will soon™ have something, but it might be a bit too involved to be included in the tests here...

oscardssmith replied:
yeah that makes sense. In that case, I think this might just need docs/news and be ready to merge (although we possibly should merge my doc PR saying "there are no docs" so that that can be backported to 1.11? not fully sure the best approach here).

kalmarek replied:
@oscardssmith Sure, merge the minimal docs PR as you please. I will not be responsive during the next week or so, proceed at your pace.

I've added news as todo entry, will do this as the last step here.

kalmarek committed:
wrap for @atomic tests in @testset

oscardssmith commented:
This looks great to me! We definitely need more tests (and review from @vtjnash to make sure that all of the default orders etc are correct).

kalmarek commented:
the default orderings are the ones from the existing cases of `@atomic` macro.

kalmarek committed:
test atomic index operations on variety of types

oscardssmith committed:
Merge branch 'master' into mk/atomic_macro_memory

kalmarek commented:
@vtjnash could you have a look at the proposed changes?
In particular the choices of orderings for the `@atomic` macro.
Do we need even more tests?

@oscardssmith what kind of docs should be written?

To me the docstring of the atomic macro is already too long, asking rather for a paragraph in the atomics part of the manual.

oscardssmith commented:
I think this does need some docs under the `@atomic ` macro, e.g.
``` 
  @atomic m[i] = new
  @atomic m[i] += addend
  @atomic :release m[i] = new
  @atomic :acquire_release m[i] += addend
```
but I agree with you that the docstring is already pretty long, but I kind of think that it is necessary to at least list the transforms that the macro does.

kalmarek committed:
document @atomic memory refs functionality

kalmarek committed:
add new @atomic functionality to language features

kalmarek committed:
update AtomicMemory docstring

kalmarek commented:
@oscardssmith @vtjnash I've updated docstrings and added NEWS entry. Please let me know if something else needs to be done/changed (i.e. please review :).

To me it seems ready to merge!

vtjnash commented on base/genericmemory.jl:
```diff
@@ -335,3 +339,94 @@ end
+    return memoryrefget(memref, order, @_boundscheck)
+end
+
+function setindex_atomic!(
+    mem::GenericMemory,
+    i::Int,
+    val,
+    order=default_access_order(mem)
+)
```
Since the index is already typically Vararg, I don't know that `order` can be optional to specify for the function (I think that is the advantage of the macro syntax is that it can supply the default):
```suggestion
setindex_atomic!(mem::GenericMemory, order::Symbol, val, inds...) = setindex_atomic!(mem, order, val, _to_linear_index(A, inds...))

function setindex_atomic!(mem::GenericMemory, order::Symbol, val, i::Int)
```
But instead of that specific former method there, you can specify that method in general following the pattern of isassigned in base/multidimensional.jl

kalmarek replied:
Can `set`/`getindex` be performed atomically with multiple indices, or rather each of the operations will be atomic, with their order unspecified? Then the name of the function would be a misnomer...

I like the suggested order of arguments (with `order` as the second one).

oscardssmith replied:
the key point here is that for AtomicMemory, the only part that needs to be atomic is the memoryrefset (it replace etc) nine of the computation of which index needs to be set needs to be atomic because we're only creating atomicity of the elements

kalmarek replied:
@oscardssmith sure, but then `setindex_atomic!(at_mem, :sequentially_consistent, [a,b], [1,2])` is not atomic, i.e. it consists of two atomic operations which can be reordered according to the order provided. That's why I said that the name would be a misnomer.

kalmarek replied:
@vtjnash Looking at base/multidimensional.jl I'm still a bit confused.
Do you want me to replace e.g. the existing `getindex_atomic` by the following?
```julia
function getindex_atomic(mem::GenericMemory, order::Symbol, i::Union{Integer,CartesianIndex}...)
    return getindex_atomic(mem, order, CartesianIndex(to_indices(mem, i)))
end
function getindex_atomic(mem::GenericMemory, order::Symbol, i::CartesianIndex)
    return getindex_atomic(mem, order, Tuple(i)...)
end
function getindex_atomic(mem::GenericMemory, order::Symbol, i::Integer...)
    idcs = to_indices(mem, i)
    S = IndexStyle(mem)
    if isa(S, IndexLinear)
        return getindex_atomic(mem, order, _to_linear_index(mem, idcs...))
    else
        return getindex_atomic(mem, order, _to_subscript_indices(mem, idcs...)...)
    end
end
function getindex_atomic(mem::GenericMemory, order::Symbol, i::Int)
    memref = memoryref(mem, i)
    return memoryrefget(memref, order, @_boundscheck)
end
```
or something yet different? Since `IndexStyle of `GenericMemory` is defined to be `IndexLinear` by abstract method, the other branch will be empty until someone actually implements some funky `GenericMemory`.

Similar methods will need to be added for `setindex_atomic!`, `setindexonce_atomic!`, `modifyindex_atomic!`, `swapindex_atomic!`, and `replaceindex_atomic!`.

Is that what you meant?

oscardssmith replied:
exactly

kalmarek replied:
3×6 methods of signature × function_name ;)

kalmarek replied:
So I wanted to do this in `genericmemory.jl`
```julia
## vararg methods generation
for (fn, args) in (
    :(getindex_atomic(mem::GenericMemory, order::Symbol, i::Int)),
    :(setindex_atomic!(mem::GenericMemory, order::Symbol, val, i::Int)),
    :(setindexonce_atomic!(mem::GenericMemory, success_order::Symbol, fail_order::Symbol, val, i::Int)),
    :(modifyindex_atomic!(mem::GenericMemory, order::Symbol, op, val, i::Int)),
    :(swapindex_atomic!(mem::GenericMemory, order::Symbol, val, i::Int)),
    :(replaceindex_atomic!(mem::GenericMemory, success_order::Symbol, fail_order::Symbol, expected, desired, i::Int,)),
)
    fn = ex.args[1]
    args = ex.args[2:end-1]

    @eval begin
        function $fn($(args...), i::Union{Integer,CartesianIndex}...)
            return $fn($(args...), CartesianIndex(to_indices($(args[1]), i)))
        end

        function $fn($(args...), i::CartesianIndex)
            return $fn($(args...), Tuple(i)...)
        end

        function $fn($(args...), i::Integer...)
            idcs = to_indices($(args[1]), i)
            S = IndexStyle($(args[1]))
            if isa(S, IndexLinear)
                return $fn($(args...), _to_linear_index($(args[1]), idcs...))
            else
                return $fn($(args...), _to_subscript_indices($(args[1]), idcs...))
            end
        end
    end
end
```
which works just fine (as julia code), but doesn't pass bootstrap. Any other ideas or should I just unroll those?

oscardssmith replied:
you could probably fix the issue by moving all of the complicated ones out of genericmemory.jl. I'm not sure why it would fail bootstrap, but `genericmemory` is pretty early in it, so lots of stuff isn't defined. Specifically, I don't think `CarteisianIndex` exists yet...

kalmarek replied:
Thanks! no it didn't, and neither did iteration etc ;)

I added this code block in `base/multidimensional.jl` in the end.

vtjnash commented on base/expr.jl:
```diff
@@ -1207,19 +1254,28 @@ function make_atomic(order, ex)
+    if is_expr(a1, :., 2)
+        a1l, a1r, op, a2 = esc(a1.args[1]), esc(a1.args[2]), esc(op), esc(a2)
+        return :(modifyproperty!($a1l, $a1r, $op, $a2, $order))
+    elseif is_expr(a1, :ref, 2)
+        x, idx, op, a2 = esc(a1.args[1]), esc(a1.args[2]), esc(op), esc(a2)
+        return :(modifyindex!($x, $idx, $op, $a2, $order))
```
I think `idx` should ideally be supporting any combination of args/kwargs here, requiring rearranging the order of all these arguments though:
```suggestion
    elseif is_expr(a1, :ref)
        x, idxs, op, a2 = esc(a1.args[1]), map(esc, a1.args[2:end]), esc(op), esc(a2)
        return :(modifyindex_atomic!($x, $order, $op, $a2, $(idxs...)))
```

kalmarek replied:
I can change the order of arguments, but I'm not sure what `z = @atomic x[1:3]` would mean, as the whole operation will not be atomic (only the individual accesses to `x` will ?)

Or do we want to have this support in macro "for future use"?

oscardssmith replied:
that won't work. the part that could is  `z = @atomic x[1, 3]`

kalmarek replied:
still, what I asked is how atomic `@atomic x[1,3]` is, if after `x[1]` is read (atomically) `x[3]` might have been modified by another thread before it is fetched to `z[2]`?

oscardssmith replied:
that's not what `x[1,3]` does. this is a matrix indexing that pulls a single entry of the matrix out atomicslly

oscardssmith replied:
`x[[1,3]]` would error

kalmarek replied:
so we want to support `GenericMemory{T} <: AbstractArray{T,1}` with 2-dimensional indexing? Or is is about already adding support for future `Memory{T,2} <: GenericMemory{T} <: AbstractArray{T,1}`?

I'm sorry but it seems that I'm very confused ;)

nsajko replied:
I think the idea is that the macro should just work for a hypothetical user package (xref [FixedSizeArrays.jl](https://github.com/JuliaArrays/FixedSizeArrays.jl)) that would add methods to `getindex_atomic` and `setindex_atomic!`.

NB: FixedSizeArrays is a nonregistered package that will presumably support `GenericMemory`, although it doesn't yet. But see one of the open PRs.

kalmarek replied:
@vtjnash I've implemented the vararg support in the macro call

vtjnash commented on base/expr.jl:
```diff
@@ -1097,29 +1097,42 @@ Mark `var` or `ex` as being performed atomically, if `ex` is a supported express
 
     @atomic a.b.x = new
-    @atomic a.b.x += addend
+    @atomic a.b.x += added
```
addend is the noun form of added (not a typo)

kalmarek commented on base/genericmemory.jl:
```diff
@@ -38,20 +38,23 @@ Memory
-    `Core.memoryrefmodify!`, and `Core.memoryrefreplace!`.
+    The access to `AtomicMemory` must be done by either using the [`@atomic`](@ref)
+    macro or the lower level interface functions: `Base.getindex_atomic`,
+    `Base.setindex_atomic!`, `Base.setindexonce_atomic!`,
+    `Base.swapindex!`, `Base.modifyindex!`, and `Base.replaceindex!`.
```
`getindex` has already implementation for `GenericMemory` and with the order of arguments it was impossible to distinguish them. Also `getindex` with multiple indices cannot(?) be performed atomically. The suffix `_atomic` here was to signify the additional property.
It mirrors the `_unsafe` convention.

We could add the `_atomic` to the latter three functions as well.

kalmarek replied:
I've added `_atomic!` suffixes to all newly implemented functions

kalmarek committed:
addend is the noun form of added (not a typo)

kalmarek committed:
add suffix `_atomic` to (swap|modify|replace)index!

kalmarek committed:
@atomic macros: add support for Vararg indices

vchuravy commented:
Would be nice to check that `Atomix.jl` still works and can be used as a backwards compatible layer to get this functionality on older Julia versions.

kalmarek commented:
@vchuravy I was actually thinking about spinning this functionality into a small package for julia-1.11, but merging into Atomix.jl would be even better. My only hesitation is that this package has not seen any activity in 2 years and was mostly developed by the brilliant **tkf** who left the community. Is it worth the effort?

vchuravy commented:
Atomix is actively used and @maleadt or I have commit privileges.

kalmarek committed:
generate vararg methods for atomic indexing fns

kalmarek committed:
[whitespace] signatures in one line when possible

kalmarek commented:
@oscardssmith @vtjnash please let me know what else need to be done here.

@nsajko, since you have access to lables I think that `needs news`, `needs docs` and `needs tests` are satisfied and could be removed.

oscardssmith committed:
Merge branch 'master' into mk/atomic_macro_memory

oscardssmith commented:
IMO this is ready to merge, but I do want @vtjnash to have a last lock

kalmarek committed:
fix doctests

oscardssmith commented:
Thanks so much for doing this!

vtjnash commented:
Awesome!
